doc/ntp/ntp-design.md

```diff
@@ -0,0 +1,1118 @@
+2. Feature should handle automatically OOB VRF changes.
+
+<!-- omit in toc -->
+#### 2.3.3.1 Server configurations: `add/del/change`
+Here is a part of `ntp.conf.j2` template that can be used to generate configured NTP servers:
+
+**Example:**
+```jinja
+{# Getting NTP global configuration -#}
+{% set global = (NTP | d({})).get('global', {}) -%}
+
+{# Adding NTP servers. We need to know if we have some pools, to set proper
+config -#}
+{% set is_pools = False %}
+{% for server in NTP_SERVER if NTP_SERVER[server].admin_state != 'disabled' and
+                               NTP_SERVER[server].resolve_as and
+                               NTP_SERVER[server].association_type -%}
+    {% set config = NTP_SERVER[server] -%}
+    {# Server options -#}
+    {% set soptions = '' -%}
+    {# Server access control options -#}
+    {% set aoptions = '' -%}
+
+    {# Authentication key -#}
+    {% if config.key -%}
+        {% set soptions = soptions ~ ' key ' ~ config.key -%}
+    {% endif -%}
+
+    {# Aggressive polling -#}
+    {% if config.iburst -%}
+        {% set soptions = soptions ~ ' iburst' -%}
+    {% endif -%}
+
+    {# Protocol version -#}
+    {% if config.version -%}
+        {% set soptions = soptions ~ ' version ' ~ config.version -%}
+    {% endif -%}
+
+    {# Check if there are any pool configured. BTW it doesn't matter what was
+    configured as "resolve_as" for pools. If they were configured with FQDN they
+    must remain like that -#}
+    {% set config_as = config.resolve_as -%}
+    {% if config.association_type == 'pool' -%}
+        {% set is_pools = True -%}
+        {% set config_as = server -%}
+    {% else -%}
+        {% set aoptions = aoptions ~ ' nopeer' -%}
+    {% endif -%}
+
+{{ config.association_type }} {{ config_as }}{{ soptions }}
+{% if global.server_role == 'disabled' %}
+restrict {{ config_as }} kod nomodify notrap noquery{{ aoptions }}
+{% endif %}
+
+{% endfor -%}
+```
+
+<!-- omit in toc -->
+#### 2.3.3.2 Global config: `change`
+The next part of `ntp.conf.j2` template can be used to manage ntp authentication:
+
+**Example:**
+```jinja
+{% set trusted_keys_arr = [] -%}
+{% for key in NTP_KEY -%}
+    {% set keydata = NTP_KEY[key] -%}
+    {% if keydata.trusted == 'yes' -%}
+        {% set trusted_keys_arr = trusted_keys_arr.append(key) -%}
+    {% endif -%}
+{% endfor %}
+
+{% if global.authentication == 'enabled' %}
+keys /etc/ntp.keys
+{% if trusted_keys_arr != [] %}
+trustedkey {{ trusted_keys_arr|join(' ') }}
+{% endif %}
+{% endif %}
+```
+
+And for handling VRF, DHCP, and NTP feature state we need to modify `ntp-sytemd-wrapper`:
+```bash
+if [ -r /etc/default/ntp ]; then
+        . /etc/default/ntp
+fi
+
+dhcp=$(/usr/local/bin/sonic-cfggen -d -v 'NTP["global"]["dhcp"]' 2> /dev/null)
+if [ -e /run/ntp.conf.dhcp ] && [ "$dhcp" = "enabled" ]; then
+        NTPD_OPTS="$NTPD_OPTS -c /run/ntp.conf.dhcp"
+fi
```

Comment:
Who creates ntp.conf.dhcp?

Should NTPD waiting for the configuration file from dhcp show up before starting?

Reply:
dhcp client.

That code already exists, I just add a condition to use it when ntp.dhcp is eanbled.
We don't need to wait for that file. If there are no dhcp ntp leases then that file won't be created.